frontend/lib/resolution/scope-queries.cpp

```diff
@@ -654,45 +656,83 @@ struct LookupHelper {
       receiverScopeHelper(receiverScopeHelper),
       checkedScopes(checkedScopes),
       result(result), foundExternBlock(foundExternBlock),
+      prevNumResults(prevNumResults),
       traceCurPath(traceCurPath), traceResult(traceResult),
       shadowedResults(shadowedResults),
       traceShadowedResults(traceShadowedResults),
       allowCached(allowCached) {
   }
 
-  bool doLookupInImportsAndUses(const Scope* scope,
-                                const ResolvedVisibilityScope* cur,
-                                UniqueString name,
-                                LookupConfig config,
-                                IdAndFlags::Flags filterFlags,
-                                const IdAndFlags::FlagSet& excludeFilter,
-                                VisibilitySymbols::ShadowScope shadowScope,
-                                bool* foundInAllContents,
-                                std::unordered_set<ID>* foundInClauses,
-                                std::unordered_set<ID>* ignoreClauses);
-
-  bool doLookupInAutoModules(const Scope* scope,
-                             UniqueString name,
-                             bool onlyInnermost,
-                             bool skipPrivateVisibilities,
-                             bool onlyMethodsFields,
-                             bool includeMethods);
+  /* There are points at which lookups check if they can be "finished".
+     E.g., if we're performing a lookup with LOOKUP_INNERMOST, we check
+     if we found something after each shadow scope level of uses/imports.
+     This method is called each time we reach such a point, and determines
+     whether we are done with the lookup.
 
-  bool doLookupEnclosingModuleName(const Scope* scope, UniqueString name);
+     For the purposes of tracking erroneous overloading cases (e.g., found
+     a function and variable with the same name at the same level), we also
+     track some mutable state (prevNumResults records the number of elements
+     in result at the point of the last call to this method).
+   */
+  bool shouldFinishLookup(const LookupResult& got, bool onlyInnermost, bool stopNonFn);
```

Comment:
I find this name somewhat confusing, reading "finish lookup" as "continue on with lookup process until it ends" rather than "stop lookup process now". The comment does make it clear though.

Reply:
I could rename it to `Stop` lookup, which might communicate the meaning better.